<a href="https://colab.research.google.com/github/JotaBlanco/TheValley/blob/main/Notebooks/01-proceso-analisis-exploratorio/01___Proceso_analisis_exploratorio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# INTRO: Proceso Análisis Exploratorio 
Introducción al EDA (Exploratory Data Analysis).

Notebook por [Javier Blanco Cordero](https://www.linkedin.com/in/javier-blanco-cordero-71373656/).

### Enlaces de interés
*   [Slides de presentación](https://docs.google.com/presentation/d/1IT6OVy9YS_sbgHF2oU9D0L--fZiyUfij7NxVhu3cAdg/edit?usp=sharing)
*   [Enlace a este notebook en Colab](https://colab.research.google.com/drive/1B37GMOCjKQ0S8NoRI6lALEW9EZtE0Hit#scrollTo=R53FGLN8jo1L)




# 01 Import
Importamos todas las librerías necesarias para este análisis ([¿No sabes lo que es una librería de Python?](https://www.quora.com/What-is-a-Python-library-and-what-can-I-use-it-for)).

In [72]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 02 Load Data
El primer paso para analizar datos es, claro, importarlos. Esto no es siempre trivial, especialmente trabajando con Colab ya que nuestro notebook está en la nube de Google y hay que conseguir hacer llegar los datos allí también. 

Vamos a ver [varias formas de importar](https://towardsdatascience.com/3-ways-to-load-csv-files-into-colab-7c14fcbdcb92) datos en Colab.


*   Importar una tabla desde Github (< 25MB)
*   Cargar (y luego importar) una tabla de nuestro local
*   Importar una tabla de Google Drive
*   Importar una tabla de Kaggle


## 0201 Importar una tabla desde GitHub

Es posible importar una tabla que guardemos en GitHub ([¿No sabes lo que es GitHub?](https://guides.github.com/activities/hello-world/)). 

Éstas deben ser pequeñas (<25MB).

1.   Las tablas deben ser pequeñas, pues GitHub no permite archivos mayores de 100MB 
2.   Localiza la tabla en un repositorio tuyo o público
3.   Pulsa en la opción RAW y copia el link de la dirección resultante
4.   Importa los datos utilizando pd.read_csv(link_raw)



### 020101 EJEMPLO 1: Covid-19
La universidad Johns Hopkins (USA) viene recopilando datos de la mayor fiabilidad posible sobre los casos y muertes confirmados por covid-19 mundialmente. Los datos son públicos y se actualizan diariamente en [su repositorio de GitHub](https://github.com/CSSEGISandData/COVID-19).

Dentro de este repositorio, hay diversos datasets, fuentes, código. Pongamos que nos interesa el dataset de casos confirmados a nivel mundial. 

¿Cómo podemos cargalo en este notebook en un Dataframe llamado df_covid?

In [73]:
# url de archivo raw de casos confirmados
url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_covid19_confirmed_global.csv'

# carga el archivo en un dataframe df_covid
df_covid = pd.read_csv(url)

df_covid.head()

Province/State Country/Region       Lat  ...  1/6/21  1/7/21  1/8/21
0            NaN    Afghanistan  33.93911  ...   53105   53207   53332
1            NaN        Albania  41.15330  ...   61008   61705   62378
2            NaN        Algeria  28.03390  ...  101120  101382  101657
3            NaN        Andorra  42.50630  ...    8348    8348    8489
4            NaN         Angola -11.20270  ...   17864   17974   18066

[5 rows x 357 columns]

## 0202 Importar una tabla local

Es posible importar una tabla que guardemos en nuestro local (en nuestro PC). 

1.   Primero tendremos que subir el archivo de nuestro local a la nube de Google 
2.   Luego generemos un buffer de los datos
3.   Finalmente los importaremos utilizando pd.read_csv()

### 020201 EJEMPLO 2: SP500
Descargad [ese dataset](https://drive.google.com/file/d/1OfTfaCTGtiS27G-uNpDsRgHqFUIaZ3XP/view?usp=sharing) de mi Google Drive personal a vuestro PC. Recoge la evolución diaria histórica del [SP&500](https://en.wikipedia.org/wiki/S%26P_500_Index). 


Ahora que está en vuestro local, ¿Cómo podríamos cargarlo a este notebook?

In [74]:
# Generamos un objeto para subir el archivo
from google.colab import files
uploaded = files.upload()

Saving sp500.csv to sp500 (1).csv


In [79]:
# Convertimos los bytes a formato tabular
import io
buffer_io = io.BytesIO(uploaded['sp500.csv'])


# Cargamos en un dataframe
df_sp500 = pd.read_csv(buffer_io)
df_sp500.head(3)

Date       Open       High        Low      Close  Adj Close  Volume
0  1927-12-30  17.660000  17.660000  17.660000  17.660000  17.660000       0
1  1928-01-03  17.760000  17.760000  17.760000  17.760000  17.760000       0
2  1928-01-04  17.719999  17.719999  17.719999  17.719999  17.719999       0

## 0203 Importar una tabla de Google Drive

Es posible importar una tabla guardada en Google Drive. Requiere:

1.   Emparejar Colab y Google Drive 
2.   Importarlo utilizando pd.read_csv()

### 020301 EJEMPLO 3: Fifa19
Descargad [este dataset](https://drive.google.com/file/d/1_qzowEIMZsLYL1VbeXb9pzSpMkPSRqWb/view?usp=sharing) con las características de los jugadores del Fifa 19 de mi Google Drive personal a vuestro pc y luego subidlo a vuestro Google Drive. 

¿Cómo podríamos cargarlo desde ahí a este notebook?

In [80]:
# Primero emparejamos Colab con Google Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [81]:
# Importamos el archivos usando read_csv()
dir_archivo = '/content/drive/My Drive/Data/Clases/fifa19.csv'
df_fifa19 = pd.read_csv(dir_archivo)
df_fifa19.head()

Unnamed: 0      ID  ... GKReflexes  Release Clause
0           0  158023  ...        8.0         €226.5M
1           1   20801  ...       11.0         €127.1M
2           2  190871  ...       11.0         €228.1M
3           3  193080  ...       94.0         €138.6M
4           4  192985  ...       13.0         €196.4M

[5 rows x 89 columns]

## 0204 EJERCICIO 1: Importar una tabla de varias formas

[Kaggle](https://www.kaggle.com/) es un sitio web imprescindible si estás aprendiendo ciencia de datos. Hay competiciones de machine learning, cursos, posts... 

Y también es un buen sitio para buscar datos, como haremos en esta ocasión. Yo me he fijado en [este](https://www.kaggle.com/miguelsxvi/air-quality-madrid-2001-2020?select=calidad_aire_madrid.csv), que contiene la evolución diaria de la concentración de diversas partículas contaminantes en Madrid. 

Leed la descripción de la página para familiarizaros con Kaggle y descargad el csv a vuestro pc. Luego, probad a cargarlo en este Notebook utilizando las 3 formas que hemos visto. (Para traerlo desde GitHub podéis utilizar [esta copia del archivo en mi GitHub personal](https://github.com/JotaBlanco/TheValley/blob/main/Data/calidad_aire_madrid.csv)). Comparad la velocidad y sencillez de cada método.

In [82]:
# DESDE GITHUB

# url de archivo raw
url = 'https://raw.githubusercontent.com/JotaBlanco/TheValley/main/Data/calidad_aire_madrid.csv'

# carga el archivo en un dataframe df_covid
df_aire = pd.read_csv(url)

df_aire.head()

estacion     fecha  magnitud  nivel
0         4  2001-1-1         1   17.0
1         4  2001-1-2         1   15.0
2         4  2001-1-3         1   15.0
3         4  2001-1-4         1   15.0
4         4  2001-1-5         1   16.0

In [83]:
# DESDE LOCAL

# Generamos un objeto para subir el archivo
from google.colab import files
uploaded = files.upload()

# Convertimos los bytes a formato tabular
import io
buffer_io = io.BytesIO(uploaded['calidad_aire_madrid.csv'])


# Cargamos en un dataframe
df_aire = pd.read_csv(buffer_io)
df_aire.head(3)

KeyError: ignored

In [84]:
# DESDE GOOGLE DRIVE

# Primero emparejamos Colab con Google Drive
from google.colab import drive
drive.mount('/content/drive')

# Importamos el archivos usando read_csv()
dir_archivo = '/content/drive/My Drive/Data/Clases/calidad_aire_madrid.csv'
df_aire = pd.read_csv(dir_archivo)
df_aire.head()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


estacion     fecha  magnitud  nivel
0         4  2001-1-1         1   17.0
1         4  2001-1-2         1   15.0
2         4  2001-1-3         1   15.0
3         4  2001-1-4         1   15.0
4         4  2001-1-5         1   16.0

# 03 Basic checks
Una vez se han importado los datos, empezamos realizando comprobaciones básicas: tamaño de la tabla, columnas presentes, etc.

El método dataframe.[info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) es muy útil para esto.

In [85]:
df_sp500.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23323 entries, 0 to 23322
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       23323 non-null  object 
 1   Open       23323 non-null  float64
 2   High       23323 non-null  float64
 3   Low        23323 non-null  float64
 4   Close      23323 non-null  float64
 5   Adj Close  23323 non-null  float64
 6   Volume     23323 non-null  int64  
dtypes: float64(5), int64(1), object(1)
memory usage: 1.2+ MB


## 0301 Vista previa
Una buena primera comprobación suele ser visualizar la tabla. 

Aquí son muy útiles los métodos dataframe.[head()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html) y dataframe.[tail()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.tail.html).

In [86]:
df_sp500.head()

Date       Open       High        Low      Close  Adj Close  Volume
0  1927-12-30  17.660000  17.660000  17.660000  17.660000  17.660000       0
1  1928-01-03  17.760000  17.760000  17.760000  17.760000  17.760000       0
2  1928-01-04  17.719999  17.719999  17.719999  17.719999  17.719999       0
3  1928-01-05  17.549999  17.549999  17.549999  17.549999  17.549999       0
4  1928-01-06  17.660000  17.660000  17.660000  17.660000  17.660000       0

In [87]:
df_sp500.head(2)

Date   Open   High    Low  Close  Adj Close  Volume
0  1927-12-30  17.66  17.66  17.66  17.66      17.66       0
1  1928-01-03  17.76  17.76  17.76  17.76      17.76       0

In [88]:
df_sp500.tail()

Date         Open  ...    Adj Close      Volume
23318  2020-10-29  3277.169922  ...  3310.110107  4903070000
23319  2020-10-30  3293.590088  ...  3269.959961  4840450000
23320  2020-11-02  3296.199951  ...  3310.239990  4310590000
23321  2020-11-03  3336.250000  ...  3369.159912  4220070000
23322  2020-11-04  3406.459961  ...  3443.439941  4783040000

[5 rows x 7 columns]

### iloc[filas,columnas]
Filtra el dataframe para quedarse con datos concretos utilizando indices.

In [89]:
df_sp500.iloc[1:3]

Date       Open       High        Low      Close  Adj Close  Volume
1  1928-01-03  17.760000  17.760000  17.760000  17.760000  17.760000       0
2  1928-01-04  17.719999  17.719999  17.719999  17.719999  17.719999       0

In [90]:
df_sp500.iloc[-5:,:5]

Date         Open         High          Low        Close
23318  2020-10-29  3277.169922  3341.050049  3259.820068  3310.110107
23319  2020-10-30  3293.590088  3304.929932  3233.939941  3269.959961
23320  2020-11-02  3296.199951  3330.139893  3279.739990  3310.239990
23321  2020-11-03  3336.250000  3389.489990  3336.250000  3369.159912
23322  2020-11-04  3406.459961  3486.250000  3405.169922  3443.439941

### loc[filas,columnas]
Filtra el dataframe para quedarse con datos concretos utilizando valores.

In [91]:
df_sp500.loc[2:5, ['Date','Open','Close']]

Date       Open      Close
2  1928-01-04  17.719999  17.719999
3  1928-01-05  17.549999  17.549999
4  1928-01-06  17.660000  17.660000
5  1928-01-09  17.500000  17.500000

In [92]:
# Admite filtros!
filtro_posterior_2018 = df_sp500['Date'] > '2018-12-31'
filtro_posterior_2018

0        False
1        False
2        False
3        False
4        False
         ...  
23318     True
23319     True
23320     True
23321     True
23322     True
Name: Date, Length: 23323, dtype: bool

In [93]:
df_sp500.loc[filtro_posterior_2018, ['Date','Open','Close']]

Date         Open        Close
22857  2019-01-02  2476.959961  2510.030029
22858  2019-01-03  2491.919922  2447.889893
22859  2019-01-04  2474.330078  2531.939941
22860  2019-01-07  2535.610107  2549.689941
22861  2019-01-08  2568.110107  2574.409912
...           ...          ...          ...
23318  2020-10-29  3277.169922  3310.110107
23319  2020-10-30  3293.590088  3269.959961
23320  2020-11-02  3296.199951  3310.239990
23321  2020-11-03  3336.250000  3369.159912
23322  2020-11-04  3406.459961  3443.439941

[466 rows x 3 columns]

## 0302 Tamaño
El tamaño de las tablas se puede comprobar con dataframe.[shape](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html), que devuelve un tupple (filas, columnas).

In [94]:
df_covid.shape

(272, 357)

In [95]:
# filas
df_covid.shape[0]

272

In [96]:
# filas
len(df_covid)

272

In [97]:
# columnas
df_covid.shape[1]

357

In [98]:
df_sp500.shape

(23323, 7)

In [99]:
df_fifa19.shape

(18207, 89)

In [100]:
df_1.shape

(844223, 4)

## 0303 Tipos de datos
Qué tipos de datos contiene nuestro dataframe?

In [101]:
df_fifa19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18207 entries, 0 to 18206
Data columns (total 89 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                18207 non-null  int64  
 1   ID                        18207 non-null  int64  
 2   Name                      18207 non-null  object 
 3   Age                       18207 non-null  int64  
 4   Photo                     18207 non-null  object 
 5   Nationality               18207 non-null  object 
 6   Flag                      18207 non-null  object 
 7   Overall                   18207 non-null  int64  
 8   Potential                 18207 non-null  int64  
 9   Club                      17966 non-null  object 
 10  Club Logo                 18207 non-null  object 
 11  Value                     18207 non-null  object 
 12  Wage                      18207 non-null  object 
 13  Special                   18207 non-null  int64  
 14  Prefer

In [102]:
df_fifa19.head(2)

Unnamed: 0      ID  ... GKReflexes  Release Clause
0           0  158023  ...        8.0         €226.5M
1           1   20801  ...       11.0         €127.1M

[2 rows x 89 columns]

In [103]:
type(df_fifa19)

pandas.core.frame.DataFrame

In [104]:
type(df_fifa19['ID'])

pandas.core.series.Series

In [105]:
type(df_fifa19['ID'][0])

numpy.int64

In [106]:
type(df_fifa19['Name'][0])

str

## 0304 Variables
El método dataframe.[columns](https://www.google.com/search?q=dataframe+columns&oq=dataframe+columns&aqs=chrome..69i57j0l5.5054j0j4&sourceid=chrome&ie=UTF-8) devuelve las columnas de nuestra tabla en orden. Es útil en lugar del método info() cuando tenemos muchas columnas, o para realizar operaciones con ellas. 

In [107]:
df_fifa19.columns

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag',
       'Overall', 'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingT

## 0305 EJEMPLO 4: Comprobaciones básicas
Hagamos las comprobaciones básicas del dataframe sobre la calidad del aire de Madrid:


*   Número de filas y columnas
*   Nombre de las columnas
*   Tipos de dato de cada columna
*   Vista previa de los primeros y últimos valores

In [108]:
# info()
df_aire.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 844223 entries, 0 to 844222
Data columns (total 4 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   estacion  844223 non-null  int64  
 1   fecha     844223 non-null  object 
 2   magnitud  844223 non-null  int64  
 3   nivel     814437 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 25.8+ MB


In [109]:
# Vista previas
display(df_aire.head())
display(df_aire.tail())

estacion     fecha  magnitud  nivel
0         4  2001-1-1         1   17.0
1         4  2001-1-2         1   15.0
2         4  2001-1-3         1   15.0
3         4  2001-1-4         1   15.0
4         4  2001-1-5         1   16.0

estacion      fecha  magnitud  nivel
844218        60  2020-3-27        14   89.0
844219        60  2020-3-28        14   76.0
844220        60  2020-3-29        14   76.0
844221        60  2020-3-30        14   82.0
844222        60  2020-3-31        14   64.0

# 04 Nulos
Los nulos son habituales y un poco molestos. Constituyen una pérdida de información y pueden impedir ciertos cálculos.

Es importante detectarlos y, cuando sea posible, corregirlos.

## 0401 Nulos por variable
Podemos ver el número de nulos por variable.

Dataframe[.isna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isna.html) es un método muy útil.

In [30]:
# Método isna()
df_fifa19.isna().head()

Unnamed: 0     ID   Name  ...  GKPositioning  GKReflexes  Release Clause
0       False  False  False  ...          False       False           False
1       False  False  False  ...          False       False           False
2       False  False  False  ...          False       False           False
3       False  False  False  ...          False       False           False
4       False  False  False  ...          False       False           False

[5 rows x 89 columns]

In [31]:
df_fifa19.isna().sum()

Unnamed: 0           0
ID                   0
Name                 0
Age                  0
Photo                0
                  ... 
GKHandling          48
GKKicking           48
GKPositioning       48
GKReflexes          48
Release Clause    1564
Length: 89, dtype: int64

In [32]:
pd.DataFrame(df_fifa19.isna().sum())

0
Unnamed: 0         0
ID                 0
Name               0
Age                0
Photo              0
...              ...
GKHandling        48
GKKicking         48
GKPositioning     48
GKReflexes        48
Release Clause  1564

[89 rows x 1 columns]

### 040101 EJEMPLO 5: Función buscadora de nulos
vamos a crear una función que, dado un dataframe, muestra una tabla con el número de nulos por variable ordenado de mayor a menor.

In [49]:
def contador_nulos(df):
  """
  Función que devuelve un dataframe indicando el número de nulos en cada
  variables. 
  """
  df_nulos = pd.DataFrame(df.isna().sum())
  df_nulos.columns = ['# Nulos']
  return df_nulos.sort_values('# Nulos', ascending=False)

In [50]:
contador_nulos(df_covid)

# Nulos
Province/State      188
Lat                   1
Long                  1
9/9/20                0
9/18/20               0
...                 ...
5/15/20               0
5/14/20               0
5/13/20               0
5/12/20               0
1/8/21                0

[357 rows x 1 columns]

In [51]:
contador_nulos(df_sp500)

# Nulos
Date             0
Open             0
High             0
Low              0
Close            0
Adj Close        0
Volume           0

In [52]:
df_contador = contador_nulos(df_fifa19)
df_contador.head()

# Nulos
Loaned From    16943
LWB             2085
LCM             2085
RS              2085
LW              2085

## 0402 Explorar Nulos
Cuando sea posible es recomendable entender la causa de los nulos. Hay algún patrón que se repite entre los nulos?

In [57]:
# La columna 'Lat' solo tenía un nulo, podemos entender por qué?
filtro_lat_nulo = df_covid['Lat'].isna()
df_covid.loc[filtro_lat_nulo, ['Province/State', 'Country/Region', 'Lat', 'Long']]

Province/State Country/Region  Lat  Long
52  Repatriated Travellers         Canada  NaN   NaN

### EJEMPLO 6: Entender nulos en 'Province/State'
En esta columna hay muchos nulos: 188. Para entender cuándo están presentes, vamos a qué paises pertenencen y a cuáles no.

In [58]:
filtro_col_nulos = df_covid['Province/State'].isna()
filtro_col_nulos

0      True
1      True
2      True
3      True
4      True
       ... 
267    True
268    True
269    True
270    True
271    True
Name: Province/State, Length: 272, dtype: bool

In [59]:
filtro_col_no_nulos = df_covid['Province/State'].isna()==False
filtro_col_no_nulos

0      False
1      False
2      False
3      False
4      False
       ...  
267    False
268    False
269    False
270    False
271    False
Name: Province/State, Length: 272, dtype: bool

In [60]:
df_covid.loc[filtro_col_nulos, 'Country/Region']

0             Afghanistan
1                 Albania
2                 Algeria
3                 Andorra
4                  Angola
              ...        
267               Vietnam
268    West Bank and Gaza
269                 Yemen
270                Zambia
271              Zimbabwe
Name: Country/Region, Length: 188, dtype: object

In [61]:
paises_provincia_nulos = df_covid.loc[filtro_col_nulos, 'Country/Region'].unique()
paises_provincia_nulos

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Austria',
       'Azerbaijan', 'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados',
       'Belarus', 'Belgium', 'Belize', 'Benin', 'Bhutan', 'Bolivia',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil', 'Brunei',
       'Bulgaria', 'Burkina Faso', 'Burma', 'Burundi', 'Cabo Verde',
       'Cambodia', 'Cameroon', 'Central African Republic', 'Chad',
       'Chile', 'Colombia', 'Comoros', 'Congo (Brazzaville)',
       'Congo (Kinshasa)', 'Costa Rica', "Cote d'Ivoire", 'Croatia',
       'Cuba', 'Cyprus', 'Czechia', 'Denmark', 'Diamond Princess',
       'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France', 'Gabon',
       'Gambia', 'Georgia', 'Germany', 'Ghana', 'Greece', 'Grenada',
       'Guatemala', 'Guinea', 'Guinea-Bissau', 'Guyana', 

In [62]:
paises_provincia_no_nulos = df_covid.loc[filtro_col_no_nulos, 'Country/Region'].unique()
paises_provincia_no_nulos

array(['Australia', 'Canada', 'China', 'Denmark', 'France', 'Netherlands',
       'United Kingdom'], dtype=object)

## 0403 Tratamiento de Nulos
Una vez hemos detectado los nulos presentes en la tabla podemos corregirlos de varias formas.

### 040301 Prescindir de Nulos
Si tenemos pocas filas con nulos y consideramos que no son muy decisivas, podemos prescindir de dichas filas con el método dataframe.[dropna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html).

In [186]:
# Tamaño del dataframe original
df_covid.shape

(272, 357)

In [187]:
# Eliminamos filas con nulos en la columna 'Lat'
df_covid_drop_null_lat = df_covid.dropna(subset=['Lat'])
df_covid_drop_null_lat.shape

(271, 357)

#### 04030101 EJEMPLO 7: Eliminar filas
En el dataframe de jugadores del Fifa, si hay ciertas filas con con gran presencia de nulos y éstas no suponen un número significativo para el total, podemos eliminarlas.

In [188]:
df_fifa19.shape

(18159, 89)

In [189]:
# Parece que hay 48 jugadores de los que no tenemos datos. Son poquitos 
# (comparado con el total de 18207) así que podemos prescindir de ellos
df_contador.head(50)

# Nulos
Loaned From             16895
LWB                      2037
LCM                      2037
RS                       2037
LW                       2037
LF                       2037
CF                       2037
RF                       2037
RW                       2037
LAM                      2037
CAM                      2037
RAM                      2037
LM                       2037
CM                       2037
LS                       2037
RCM                      2037
RM                       2037
LDM                      2037
CDM                      2037
RDM                      2037
RWB                      2037
LB                       2037
LCB                      2037
CB                       2037
RCB                      2037
RB                       2037
ST                       2037
Release Clause           1516
Joined                   1505
Club                      241
Contract Valid Until      241
Jersey Number              12
Position                   12
ShotPower                   0
Aggression                  0
LongShots                   0
Strength                    0
Stamina                     0
Jumping                     0
Weight                      0
Balance                     0
Reactions                   0
Interceptions               0
SprintSpeed                 0
Acceleration                0
BallControl                 0
Agility                     0
Vision                      0
Positioning                 0
FKAccuracy                  0

In [190]:
# Eliminamos aquellas filas que no tienen informado el peso del jugardor
df_fifa19.dropna(axis=0, subset=['Weight'], inplace=True)
df_contador = contador_nulos(df_fifa19)
df_contador.head(50)

# Nulos
Loaned From             16895
LWB                      2037
LCM                      2037
RS                       2037
LW                       2037
LF                       2037
CF                       2037
RF                       2037
RW                       2037
LAM                      2037
CAM                      2037
RAM                      2037
LM                       2037
CM                       2037
LS                       2037
RCM                      2037
RM                       2037
LDM                      2037
CDM                      2037
RDM                      2037
RWB                      2037
LB                       2037
LCB                      2037
CB                       2037
RCB                      2037
RB                       2037
ST                       2037
Release Clause           1516
Joined                   1505
Club                      241
Contract Valid Until      241
Jersey Number              12
Position                   12
ShotPower                   0
Aggression                  0
LongShots                   0
Strength                    0
Stamina                     0
Jumping                     0
Weight                      0
Balance                     0
Reactions                   0
Interceptions               0
SprintSpeed                 0
Acceleration                0
BallControl                 0
Agility                     0
Vision                      0
Positioning                 0
FKAccuracy                  0

In [191]:
df_fifa19.shape

(18159, 89)

### 040302 Rellenar Nulos con Valores
Podemos sustituir los nulos con valores concretos: 0, la media o mediana, etc.

Para esto podemos utilizar el método dataframe.[fillna()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html#pandas.DataFrame.fillna).

In [192]:
# Rellenar nulos con 0s
df_fifa19['Release Clause'] = df_fifa19['Release Clause'].fillna(0)

# 05 Limpieza de datos
Es habitual que algunos de los datos se encuentren en formatos no apropiados.


## 0501 Limpieza de Formatos
A veces los datos se encuentran en formatos no adecuados. Un formato correspondiente con el tipo de dato permitirá realizar ciertas operaciones con mayor facilidad.

Para esto nos serán de utilidad los métodos:

*   [astype()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.astype.html)
*   [to_datetime()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html)

In [193]:
df_sp500.loc[:5,'Date']

0   1927-12-30
1   1928-01-03
2   1928-01-04
3   1928-01-05
4   1928-01-06
5   1928-01-09
Name: Date, dtype: datetime64[ns]

In [194]:
df_sp500.loc[:5,'Date'][0]

Timestamp('1927-12-30 00:00:00')

In [195]:
type(df_sp500.loc[:5,'Date'][0])

pandas._libs.tslibs.timestamps.Timestamp

In [196]:
# Cambiamos el formato de la columna 'Date'
df_sp500['Date'] = pd.to_datetime(
    df_sp500['Date'], format='%Y-%m-%d', errors='ignore')

## 0502 Eliminación de variables
Se pueden eliminar columnas o filas no necesarias con el método dataframe.[drop()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html).

In [197]:
df_fifa19.head(3)

Unnamed: 0      ID  ... GKReflexes  Release Clause
0           0  158023  ...        8.0         €226.5M
1           1   20801  ...       11.0         €127.1M
2           2  190871  ...       11.0         €228.1M

[3 rows x 89 columns]

In [198]:
df_fifa19.columns

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag',
       'Overall', 'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingT

In [199]:
cols_to_drop = ['Unnamed: 0', 'ID', 'Photo', 'Flag', 'Club Logo', 
                'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW', 'LAM', 'CAM', 
                'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM', 'CDM', 
                'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB']

df_fifa19_clean = df_fifa19.drop(cols_to_drop, axis=1)

In [200]:
df_fifa19.shape, df_fifa19_clean.shape

((18159, 89), (18159, 58))

# 06 Transformación de variables
A veces querremos crear nuevas variables o modificar variables existentes de manera un poco más personalizada/compleja que lo visto hasta ahora. Cómo lo hacemos?

Para cambiar los valores en una columna debemos tratar de evitar iterar sobre las filas del dataframe. Son procesos lentos que llevarán mucho tiempo para dataframes grandes. El orden (atendiendo a la eficiencia computacional) en que debemos intentar las operaciones es:



1.   Vectorización
2.   List comprehensions
3.   Dataframe.[apply()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.apply.html)
4.   Dataframe.[itertuples()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.itertuples.html), dataframe.[iteritems()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iteritems.html#pandas.DataFrame.iteritems), dataframe.[iterrows()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iterrows.html#pandas.DataFrame.iterrows)



## 0601 Iteracion
Aunque sea el menos eficiente, iterar sobre el dataframe es lo más sencillo conceptualmente. Veamos cómo podemos modificar una variable así.

In [180]:
for index, row in df_fifa19_clean.head().iterrows():
  print('Index: ', index)
  print('Row: ', row)

Index:  0
Row:  Name                              L. Messi
Age                                     31
Nationality                      Argentina
Overall                                 94
Potential                               94
Club                          FC Barcelona
Value                              €110.5M
Wage                                 €565K
Special                               2202
Preferred Foot                        Left
International Reputation                 5
Weak Foot                                4
Skill Moves                              4
Work Rate                   Medium/ Medium
Body Type                            Messi
Real Face                              Yes
Position                                RF
Jersey Number                           10
Joined                         Jul 1, 2004
Loaned From                            NaN
Contract Valid Until                  2021
Height                                 5'7
Weight                              15

In [181]:
for index, row in df_fifa19_clean.head().iterrows():
  print('Index: ', index)
  print('Peso: ', row['Weight'])

Index:  0
Peso:  159lbs
Index:  1
Peso:  183lbs
Index:  2
Peso:  150lbs
Index:  3
Peso:  168lbs
Index:  4
Peso:  154lbs


In [183]:
for index, row in df_fifa19_clean.iterrows():
  
  # Si el string contine 'lbs' al final del mismo
  if row['Weight'][-3:] == 'lbs':

    # Nos quedamos con la parte anterior (el peso) con formato numérico (float)
    peso_libras = float(row['Weight'][:-3])

    # Convertimos el peso en kgs
    peso_kgs = peso_libras*0.453592

    # Populamos la columna
    df_fifa19_clean.loc[index, 'Weight_cl'] = peso_kgs
  
  else:
    # Si no contenía 'lbs' al final, lo dejamos como nulo
    df_fifa19_clean.loc[index, 'Weight_cl'] = np.NaN

df_fifa19_clean.head()

Name  Age Nationality  ...  GKReflexes  Release Clause  Weight_cl
0           L. Messi   31   Argentina  ...         8.0         €226.5M  72.121128
1  Cristiano Ronaldo   33    Portugal  ...        11.0         €127.1M  83.007336
2          Neymar Jr   26      Brazil  ...        11.0         €228.1M  68.038800
3             De Gea   27       Spain  ...        94.0         €138.6M  76.203456
4       K. De Bruyne   27     Belgium  ...        13.0         €196.4M  69.853168

[5 rows x 59 columns]

In [185]:
# Falla :S por la presencia de nulos, corregimos tal como habíamos aprendido
df_fifa19_clean['Weight_cl'].isna().sum()

0

In [159]:
df_fifa19_clean.loc[13232:13240,:]

Name  Age Nationality  ...  GKPositioning  GKReflexes Release Clause
13232     A. Rizzo   21       Italy  ...           13.0        10.0          €875K
13233    L. Thomas   19     England  ...            7.0         8.0            NaN
13234  R. Schlegel   21   Argentina  ...            7.0        15.0          €808K
13235   D. Montoya   22    Colombia  ...            9.0         8.0          €543K
13236   J. McNulty   33    Scotland  ...            NaN         NaN            NaN
13237   J. Barrera   29   Nicaragua  ...            NaN         NaN            NaN
13238     J. Stead   35     England  ...            NaN         NaN            NaN
13239  A. Semprini   20       Italy  ...            NaN         NaN            NaN
13240   R. Bingham   24     England  ...            NaN         NaN            NaN

[9 rows x 58 columns]

## 0503 Creación de nuevas variables
Del mismo modo, a veces puede ser útil crear nuevas variables.

In [130]:
# Nueva variable año
df_sp500['Year'] = df_sp500['Date'].dt.year
df_sp500.head()

Date       Open       High  ...  Adj Close  Volume  Year
0 1927-12-30  17.660000  17.660000  ...  17.660000       0  1927
1 1928-01-03  17.760000  17.760000  ...  17.760000       0  1928
2 1928-01-04  17.719999  17.719999  ...  17.719999       0  1928
3 1928-01-05  17.549999  17.549999  ...  17.549999       0  1928
4 1928-01-06  17.660000  17.660000  ...  17.660000       0  1928

[5 rows x 8 columns]

In [131]:
# Nueva variable volatilidad diaria
df_sp500['Volatilidad Diaria'] = (df_sp500['High']-df_sp500['Low'])/df_sp500['Close']
df_sp500.head()

Date       Open       High  ...  Volume  Year  Volatilidad Diaria
0 1927-12-30  17.660000  17.660000  ...       0  1927                 0.0
1 1928-01-03  17.760000  17.760000  ...       0  1928                 0.0
2 1928-01-04  17.719999  17.719999  ...       0  1928                 0.0
3 1928-01-05  17.549999  17.549999  ...       0  1928                 0.0
4 1928-01-06  17.660000  17.660000  ...       0  1928                 0.0

[5 rows x 9 columns]

In [132]:
df_sp500.tail()

Date         Open  ...  Year  Volatilidad Diaria
23318 2020-10-29  3277.169922  ...  2020            0.024540
23319 2020-10-30  3293.590088  ...  2020            0.021710
23320 2020-11-02  3296.199951  ...  2020            0.015225
23321 2020-11-03  3336.250000  ...  2020            0.015802
23322 2020-11-04  3406.459961  ...  2020            0.023546

[5 rows x 9 columns]

In [134]:
# Nueva variable movilidad diaria disponible
filtro_movilidad_diaria_disponible = df_sp500['High'] != df_sp500['Low']
filtro_movilidad_diaria_disponible

0        False
1        False
2        False
3        False
4        False
         ...  
23318     True
23319     True
23320     True
23321     True
23322     True
Length: 23323, dtype: bool

In [135]:
df_sp500['Movilidad Diaria Disponible'] = filtro_movilidad_diaria_disponible.astype(int)
df_sp500['Movilidad Diaria Disponible']

0        0
1        0
2        0
3        0
4        0
        ..
23318    1
23319    1
23320    1
23321    1
23322    1
Name: Movilidad Diaria Disponible, Length: 23323, dtype: int64

## 0504 Limpieza de valores
Algunas de las columnas pueden requerir transformaciones para obtener un formato deseado.

In [137]:
# Weight
df_fifa19_clean['Weight']

0        159lbs
1        183lbs
2        150lbs
3        168lbs
4        154lbs
          ...  
18202    134lbs
18203    170lbs
18204    148lbs
18205    154lbs
18206    176lbs
Name: Weight, Length: 18207, dtype: object

# 04 Distribución de variables


In [41]:
df_fifa19.head(2)

Unnamed: 0      ID  ... GKReflexes  Release Clause
0           0  158023  ...        8.0         €226.5M
1           1   20801  ...       11.0         €127.1M

[2 rows x 89 columns]

In [42]:
df_fifa19['Age'], df_fifa19['Overall']

(0        31
 1        33
 2        26
 3        27
 4        27
          ..
 18202    19
 18203    19
 18204    16
 18205    17
 18206    16
 Name: Age, Length: 18207, dtype: int64, 0        94
 1        94
 2        92
 3        91
 4        91
          ..
 18202    47
 18203    47
 18204    47
 18205    47
 18206    46
 Name: Overall, Length: 18207, dtype: int64)

## 0401 Variables numéricas
Variables que adoptan valores numéricos.

In [43]:
df_fifa19.iloc[:,2:].describe()

Age       Overall  ...  GKPositioning    GKReflexes
count  18207.000000  18207.000000  ...   18159.000000  18159.000000
mean      25.122206     66.238699  ...      16.388898     16.710887
std        4.669943      6.908930  ...      17.034669     17.955119
min       16.000000     46.000000  ...       1.000000      1.000000
25%       21.000000     62.000000  ...       8.000000      8.000000
50%       25.000000     66.000000  ...      11.000000     11.000000
75%       28.000000     71.000000  ...      14.000000     14.000000
max       45.000000     94.000000  ...      90.000000     94.000000

[8 rows x 42 columns]

## Variables categóricas
Aquellas que pueden adoptar valores finitos.

In [44]:
df_fifa19['Nationality'].value_counts()

England                 1662
Germany                 1198
Spain                   1072
Argentina                937
France                   914
                        ... 
United Arab Emirates       1
Palestine                  1
Puerto Rico                1
New Caledonia              1
Rwanda                     1
Name: Nationality, Length: 164, dtype: int64

In [45]:
df_fifa19['Preferred Foot'].value_counts()

Right    13948
Left      4211
Name: Preferred Foot, dtype: int64

In [46]:
df_fifa19.head()

Unnamed: 0      ID  ... GKReflexes  Release Clause
0           0  158023  ...        8.0         €226.5M
1           1   20801  ...       11.0         €127.1M
2           2  190871  ...       11.0         €228.1M
3           3  193080  ...       94.0         €138.6M
4           4  192985  ...       13.0         €196.4M

[5 rows x 89 columns]